### ISSUE TITLE
I got this error ”keyserver receive failed: Server indicated a failure“

### ISSUE THREAD

**Issue by BboyFate** [docker-library/python#597]
Where did I go wrong？
Please look at it. Thank you！
I used this template [Dockerfile-debian.template](https://github.com/docker-library/python/blob/master/Dockerfile-debian.template)

```
+ mktemp -d
+ export GNUPGHOME=/tmp/tmp.oMdzi8CVS1
+ gpg --batch --keyserver ha.pool.sks-keyservers.net --recv-keys E3FF2839C048B25C084DEBE9B26995E310250568
gpg: keybox '/tmp/tmp.oMdzi8CVS1/pubring.kbx' created
gpg: keyserver receive failed: Server indicated a failure
```

**Comment by wglambert** [comment]
https://github.com/docker-library/official-images/issues/4252#issuecomment-381783035
>TLDR; GPG servers are extremely flaky . . .

This is what we do for GPG issues https://github.com/docker-library/official-images/issues/4252#issuecomment-405749749

You can keep running the build and it will eventually succeed since `ha.pool.sks-keyservers.net` is a round-robin pool
https://github.com/docker-library/httpd/issues/66#issuecomment-316832441

Or you could add more servers to try https://github.com/docker-library/php/issues/586#issue-296627087

https://github.com/docker-library/tomcat/issues/87#issuecomment-344452350
>My own preferred ordering would be the following:

- `ha.pool.sks-keyservers.net`
- `pgp.mit.edu`
- `hkp://p80.pool.sks-keyservers.net:80`
- `hkp://keyserver.ubuntu.com:80`
- `keyserver.pgp.com`

**Comment by christianbundy** [comment]
I've tried all of the above servers and none of them seem to have `E3FF2839C048B25C084DEBE9B26995E310250568`. Can anyone else receive that key?

**Comment by tianon** [comment]
The email address on that key is verified on https://keys.openpgp.org, so you can use `hkps://keys.openpgp.org` for sure (see #599).

**Comment by BboyFate** [comment]
> [docker-library/official-images#4252 (comment)](https://github.com/docker-library/official-images/issues/4252#issuecomment-381783035)
> 
> > TLDR; GPG servers are extremely flaky . . .
> 
> This is what we do for GPG issues [docker-library/official-images#4252 (comment)](https://github.com/docker-library/official-images/issues/4252#issuecomment-405749749)
> 
> You can keep running the build and it will eventually succeed since `ha.pool.sks-keyservers.net` is a round-robin pool
> [docker-library/httpd#66 (comment)](https://github.com/docker-library/httpd/issues/66#issuecomment-316832441)
> 
> Or you could add more servers to try [docker-library/php#586 (comment)](https://github.com/docker-library/php/issues/586#issue-296627087)
> 
> [docker-library/tomcat#87 (comment)](https://github.com/docker-library/tomcat/issues/87#issuecomment-344452350)
> 
> > My own preferred ordering would be the following:
> 
> * `ha.pool.sks-keyservers.net`
> * `pgp.mit.edu`
> * `hkp://p80.pool.sks-keyservers.net:80`
> * `hkp://keyserver.ubuntu.com:80`
> * `keyserver.pgp.com`

Thank you very much!
I used the second method eventually succeed:
```
·
·
·
&& wget -O python.tar.xz "https://www.python.org/ftp/python/${PYTHON_VERSION%%[a-z]*}/Python-$PYTHON_VERSION.tar.xz" \
&& wget -O python.tar.xz.asc "https://www.python.org/ftp/python/${PYTHON_VERSION%%[a-z]*}/Python-$PYTHON_VERSION.tar.xz.asc" \
&& export GNUPGHOME="$(mktemp -d)" \
&& for GPG_KEY in \
	 E3FF2839C048B25C084DEBE9B26995E310250568 \
	 ; do \
      gpg --batch --keyserver ha.pool.sks-keyservers.net --recv-keys "$GPG_KEY" || \
      gpg --batch --keyserver pgp.mit.edu --recv-keys "$GPG_KEY" || \
      gpg --batch --keyserver hkp://p80.pool.sks-keyservers.net:80 --recv-keys "$GPG_KEY" || \
      gpg --batch --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys "$GPG_KEY" || \
      gpg --batch --keyserver keyserver.pgp.com --recv-keys "$GPG_KEY" || \
      gpg --batch --keyserver hkp://pgp.mit.edu:80 --recv-keys "$GPG_KEY" \
      gpg --batch --keyserver hkp://ipv4.pool.sks-keyservers.net --recv-keys "$GPG_KEY" ; \
    done \
&& gpg --batch --verify python.tar.xz.asc python.tar.xz \
·
·
·
```

**Comment by BboyFate** [comment]
> The email address on that key is verified on https://keys.openpgp.org, so you can use `hkps://keys.openpgp.org` for sure (see #599).

yep, but i'm add more servers

**Comment by christianbundy** [comment]
Sorry, am I doing this right?

```console
$ gpg --batch --keyserver https://keys.openpgp.org --recv-keys E3FF2839C048B25C084DEBE9B26995E310250568
gpg: no valid OpenPGP data found.
gpg: Total number processed: 0
```

**Comment by tianon** [comment]
You want to use `hkps`, not `https` (as in #599).

**Comment by christianbundy** [comment]
Thanks for the feedback. That's what I tried at first, but that gives the same error as the other keyservers:

```console
$ gpg --batch --keyserver hkps://keys.openpgp.org --recv-keys E3FF2839C048B25C084DEBE9B26995E310250568
gpg: keyserver receive failed: Server indicated a failure
```

Could you confirm that this command works for you? (Or am I mistyping something?)

**Comment by tianon** [comment]
Yep, I can confirm that exact command works fine for me:

```console
$ docker run -it --rm --dns 1.1.1.1 buildpack-deps:buster-curl gpg --batch --keyserver hkps://keys.openpgp.org --recv-keys E3FF2839C048B25C084DEBE9B26995E310250568
gpg: directory '/root/.gnupg' created
gpg: keybox '/root/.gnupg/pubring.kbx' created
gpg: /root/.gnupg/trustdb.gpg: trustdb created
gpg: key B26995E310250568: public key "\xc5\x81ukasz Langa (GPG langa.pl) <lukasz@langa.pl>" imported
gpg: Total number processed: 1
gpg:               imported: 1
```

**Comment by christianbundy** [comment]
Oof. Can confirm that it works in Docker, but not MacOS. Sorry for the noise and thanks for the help.

**Comment by BboyFate** [comment]
> Oof. Can confirm that it works in Docker, but not MacOS. Sorry for the noise and thanks for the help.

You can try those keyservers :
Use cyclely
```
for GPG_KEY in \
	 E3FF2839C048B25C084DEBE9B26995E310250568 \
	 ; do \
      gpg --batch --keyserver ha.pool.sks-keyservers.net --recv-keys "$GPG_KEY" || \
      gpg --batch --keyserver pgp.mit.edu --recv-keys "$GPG_KEY" || \
      gpg --batch --keyserver hkp://p80.pool.sks-keyservers.net:80 --recv-keys "$GPG_KEY" || \
      gpg --batch --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys "$GPG_KEY" || \
      gpg --batch --keyserver keyserver.pgp.com --recv-keys "$GPG_KEY" || \
      gpg --batch --keyserver hkp://pgp.mit.edu:80 --recv-keys "$GPG_KEY" \
      gpg --batch --keyserver hkp://ipv4.pool.sks-keyservers.net --recv-keys "$GPG_KEY" ; \
    done \
```

**Comment by ernstki** [comment]
For those (still) finding themselves here trying to verify a tarball from python.org/downloads and doing a web search for the key ID E3FF2839C048B25C084DEBE9B26995E310250568…

Python 3.8 releases, for example, were signed with this key.

As seen above — when you finally contact a GPG/PGP keyserver that works — the new key ID appears to be `B26995E310250568`. This new key ID is mentioned [here](https://test.python.org/downloads/metadata/pgp/), if you're like me and want to verify these things.

I always feel like I'm missing something when a project provides a PGP/GPG key with no obvious instructions for how to verify it, but [getting that right](https://keepassxc.org/download) is hard, I guess. Meanwhile, Python has moved on to [some other verification method](https://www.python.org/downloads/metadata/sigstore/), making this practically a moot point.
